auth/authzserver/provider.go

```diff
@@ -0,0 +1,261 @@
+)
+
+const (
+	ClientIDClaim = "client_id"
+	UserIDClaim   = "user_info"
+	ScopeClaim    = "scp"
+	KeyIDClaim    = "key_id"
+)
+
+type Provider struct {
+	fosite.OAuth2Provider
+	cfg       config.AuthorizationServer
+	publicKey []rsa.PublicKey
+	keySet    jwk.Set
+}
+
+func (p Provider) PublicKeys() []rsa.PublicKey {
+	return p.publicKey
+}
+
+func (p Provider) KeySet() jwk.Set {
+	return p.keySet
+}
+
+// NewJWTSessionToken is a helper function for creating a new session.
+func (p Provider) NewJWTSessionToken(subject, appID, issuer, audience string, userInfoClaims *service.UserInfoResponse) *fositeOAuth2.JWTSession {
+	key, found := p.keySet.Get(0)
+	keyID := ""
+	if found {
+		keyID = key.KeyID()
+	}
+
+	return &fositeOAuth2.JWTSession{
+		JWTClaims: &jwt.JWTClaims{
+			Audience:  []string{audience},
+			Issuer:    issuer,
+			Subject:   subject,
+			ExpiresAt: time.Now().Add(p.cfg.AccessTokenLifespan.Duration),
+			IssuedAt:  time.Now(),
+			Extra: map[string]interface{}{
+				ClientIDClaim: appID,
+				UserIDClaim:   userInfoClaims,
+			},
+		},
+		JWTHeader: &jwt.Headers{
+			Extra: map[string]interface{}{
+				KeyIDClaim: keyID,
+			},
+		},
+	}
+}
+
+func findPublicKeyForTokenOrFirst(ctx context.Context, t *jwtgo.Token, publicKeys jwk.Set) (*rsa.PublicKey, error) {
+	if _, ok := t.Method.(*jwtgo.SigningMethodRSA); !ok {
+		return nil, fmt.Errorf("unexpected signing method: %v", t.Header["alg"])
+	}
+
+	if publicKeys.Len() == 0 {
+		return nil, fmt.Errorf("no keys exist to match")
+	}
+
+	publicKey := &rsa.PublicKey{}
+	k, _ := publicKeys.Get(0)
+	if err := k.Raw(publicKey); err != nil {
+		return nil, err
+	}
+
+	if keyID, found := t.Header[KeyIDClaim]; !found {
+		return publicKey, nil
+	} else if key, found := publicKeys.LookupKeyID(keyID.(string)); !found {
+		return publicKey, nil
+	} else if err := key.Raw(publicKey); err != nil {
+		logger.Errorf(ctx, "Failed to load public key from key [%v]. Will default to the first key. Error: %v", keyID)
+		return publicKey, nil
+	}
+
+	return publicKey, nil
+}
+
+func (p Provider) ValidateAccessToken(ctx context.Context, expectedAudience, tokenStr string) (interfaces.IdentityContext, error) {
+	// Parse and validate the token.
+	parsedToken, err := jwtgo.Parse(tokenStr, func(t *jwtgo.Token) (interface{}, error) {
+		return findPublicKeyForTokenOrFirst(ctx, t, p.KeySet())
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	if !parsedToken.Valid {
+		return nil, fmt.Errorf("parsed token is invalid")
+	}
+
+	claimsRaw := parsedToken.Claims.(jwtgo.MapClaims)
+	return verifyClaims(expectedAudience, claimsRaw)
+}
+
+func verifyClaims(expectedAudience string, claimsRaw map[string]interface{}) (interfaces.IdentityContext, error) {
+	claims := jwtx.ParseMapStringInterfaceClaims(claimsRaw)
+	if len(claims.Audience) != 1 {
+		return nil, fmt.Errorf("expected exactly one granted audience. found [%v]", len(claims.Audience))
+	}
+
+	if claims.Audience[0] != expectedAudience {
```

Comment:
```suggestion
	if !strings.Contains(expectedAudience, claims.Audience[0]) {
```
It looks the code will always prepend `http/https` to the url but in our case the `audience[0]` does not have that prefix

Reply:
What do you use as an Authorization Server? Okta? KeyCloak? if so, you can configure them to issue tokens with a specific audience...